File: src/geo-frontend/src/assets/js/websocket/modules/UploadJobModule.js
/**
 * Upload Job WebSocket module.
 * Handles all upload job related realtime events for the import queue.
 */

import {BaseModule} from './BaseModule.js';

export class UploadJobModule extends BaseModule {
    constructor(store) {
        super(store);
        this.moduleName = 'upload_job';
    }

    /**
     * Initialize the upload job module
     */
    initialize() {
        super.initialize();

        // Handle status updates
        this.subscribe('status_updated', (data) => {
            console.log('Upload job status updated:', data);
            this.handleStatusUpdate(data);
        });

        // Handle upload job completion
        this.subscribe('completed', (data) => {
            console.log('Upload job completed:', data);
            // Request refresh of import queue to get updated data
            this.socket.requestRefresh('import_queue');
        });

        // Handle upload job failure
        this.subscribe('failed', (data) => {
            console.log('Upload job failed:', data);
            // Request refresh of import queue to get updated data
            this.socket.requestRefresh('import_queue');
        });
    }

    /**
     * Handle status update events
     * @param {Object} data - The status update data
     */
    handleStatusUpdate(data) {
        let updates = {
            processing: data.status === 'processing',
            processing_failed: data.status === 'failed'
        };

        // Handle completed status - need to get the actual feature count from the server
        if (data.status === 'completed') {
            updates.processing = false;
            updates.processing_failed = false;
            // Request a refresh of import queue to get the updated item with correct feature count
            this.socket.requestRefresh('import_queue');
            return;
        }

        // For processing status, set feature_count to -1 to indicate processing
        if (data.status === 'processing') {
            updates.feature_count = -1;
        }

        // Update the specific item in the store using import_queue_id
        this.store.dispatch('updateImportQueueItem', {
            id: data.import_queue_id,
            updates: updates
        });
    }

    /**
     * Cleanup the upload job module
     */
    cleanup() {
        super.cleanup();
        console.log('Upload job module cleaned up');
    }
}

File: src/geo-backend/geo_lib/websocket/modules/upload_job_module.py
"""
Upload job WebSocket module.
Handles upload job processing events (status updates, completion, etc.).
"""

from geo_lib.websocket.base_module import BaseWebSocketModule
from geo_lib.logging.console import get_websocket_logger

logger = get_websocket_logger()


class UploadJobModule(BaseWebSocketModule):
    """WebSocket module for upload job processing events."""
    
    @property
    def module_name(self) -> str:
        return "upload_job"
    
    async def handle_message(self, message_type: str, data: dict) -> None:
        """Handle incoming messages for upload job module."""
        if message_type == 'refresh':
            # Upload job module doesn't need to send initial state
            # The import queue module handles that
            pass
        else:
            logger.warning(f"Unknown message type for upload_job module: {message_type}")
    
    async def send_initial_state(self) -> None:
        """Upload job module doesn't send initial state - import queue module handles that."""
        # Upload job module doesn't need to send initial state
        # The import queue module handles the initial state for the UI
        pass
    
    # WebSocket event handlers for upload job events
    async def status_updated(self, event):
        """Handle upload job status updates."""
        await self.send_to_client('status_updated', event['data'])
    
    async def completed(self, event):
        """Handle upload job completion."""
        await self.send_to_client('completed', event['data'])
    
    async def failed(self, event):
        """Handle upload job failure."""
        await self.send_to_client('failed', event['data'])
